pontoon/batch/tests/test_views.py

```diff
@@ -11,56 +79,151 @@ def test_batch_edit_translations_no_user(client):
-            'action': 'reject',
-            'locale': locale_a.code,
-            'entities': entity_a.id,
-        },
-        HTTP_X_REQUESTED_WITH='XMLHttpRequest',
+    response = batch_action(
+        action='reject',
+        locale=locale_a.code,
+        entities=entity_a.id,
     )
+
     assert response.status_code == 403
     assert 'Forbidden' in response.content
+
+
+@pytest.mark.django_db
+def test_batch_approve_valid_translations(
+    batch_action,
+    member,
+    translation_dtd_unapproved,
+):
+    """
+    Approve translations without errors.
+    """
+    response = batch_action(
+        admin=True,
+        action='approve',
+        locale=translation_dtd_unapproved.locale.code,
+        entities=translation_dtd_unapproved.entity.pk,
+    )
+    assert response.json() == {
+        'count': 1,
+        'invalid_translation_count': 0,
+    }
+
+    translation_dtd_unapproved.refresh_from_db()
+    assert translation_dtd_unapproved.approved
+
+
+@pytest.mark.django_db
+def test_batch_approve_invalid_translations(
+    batch_action,
+    member,
+    translation_dtd_invalid_unapproved,
+):
+    """
+    Translations with errors can't be approved.
+    """
+
+    response = batch_action(
+        admin=True,
+        action='approve',
+        locale=translation_dtd_invalid_unapproved.locale.code,
+        entities=translation_dtd_invalid_unapproved.entity.pk
+    )
+
+    assert response.json() == {
+        'count': 0,
+        'invalid_translation_count': 1,
+    }
+
+    translation_dtd_invalid_unapproved.refresh_from_db()
+
+    assert not translation_dtd_invalid_unapproved.approved
+
+
+@pytest.mark.django_db
+def test_batch_find_and_replace_valid_translations(
+    batch_action,
+    member,
+    translation_dtd_unapproved,
+):
+    response = batch_action(
+        admin=True,
+        action='replace',
+        locale=translation_dtd_unapproved.locale.code,
+        entities=translation_dtd_unapproved.entity.pk,
+        find='Translation',
+        replace='Replaced translation',
+    )
+
+    assert response.json() == {
+        'count': 1,
+        'invalid_translation_count': 0,
+    }
+
+    new_translation = translation_dtd_unapproved.entity.translation_set.last()
+
+    assert new_translation.string == 'Test Replaced translation'
+    assert new_translation.approved
+
+
+@pytest.mark.django_db
+def test_batch_find_and_replace_invalid_translations(
+    batch_action,
+    member,
+    translation_dtd_unapproved,
+):
+    """
+    The `find & replace` action can't produce invalid translations.
+    """
+    response = batch_action(
+        admin=True,
+        action='replace',
+        locale=translation_dtd_unapproved.locale.code,
+        entities=translation_dtd_unapproved.entity.pk,
+        find='Translation',
+        replace='%$#%>',
+    )
+
+    assert response.json() == {
+        'count': 0,
+        'invalid_translation_count': 1,
+    }
+
+    translation_dtd_unapproved.refresh_from_db()
```

Comment:
The string of `translation_dtd_unapproved` would never change. I'd use the `new_translation` trick from above here, too.